virtool/users/transforms.py

```diff
@@ -0,0 +1,97 @@
+import asyncio
+from typing import List, Dict, Any, Union, Optional, TYPE_CHECKING
+
+from sqlalchemy import select
+from sqlalchemy.ext.asyncio import AsyncEngine
+
+from virtool.data.topg import both_transactions
+from virtool.groups.pg import SQLGroup
+from virtool.groups.utils import merge_group_permissions
+from virtool.mongo.transforms import AbstractTransform
+from virtool.types import Document
+from virtool.utils import get_safely
+
+if TYPE_CHECKING:
+    from virtool.mongo.core import Mongo
+
+
+class AttachPermissionsTransform(AbstractTransform):
+    """
+    Attaches more complete task data to a document with a `task.id` field.
+    """
+
+    def __init__(self, pg: AsyncEngine, mongo: "Mongo"):
+        self._pg = pg
+        self._mongo = mongo
+
+    async def attach_one(self, document, prepared):
+        return {
+            **document,
+            "permissions": merge_group_permissions(
+                [prepared[group_id] for group_id in self.get_group_ids(document)]
+            ),
+        }
+
+    async def attach_many(
+        self, documents: List[Document], prepared: Dict[str, any]
+    ) -> List[Document]:
+        attached = []
+
+        for document in documents:
+            attached.append(
+                {
+                    **document,
+                    "permissions": merge_group_permissions(
+                        [
+                            prepared[group_id]
+                            for group_id in self.get_group_ids(document)
+                        ]
+                    ),
+                }
+            )
+
+        return attached
+
+    async def prepare_one(self, document) -> Optional[Document]:
+        group_ids = self.get_group_ids(document)
+        return await self.get_groups_from_ids(list(group_ids))
+
+    async def prepare_many(
+        self, documents: List[Document]
```

Comment:
```suggestion
        self, documents: list[Document]
```
You are inconsistenly using type names list `Dict` and `dict`.

We should try to move to the new style `dict` and `list` instead of `List` and `Dict`. Instead of `Optional` and `Union`, you can also use `str | None` or `int | str`. I am not stuck on every module being updated, but new work should at least be consistent with what is in the module already.